docs/architecture/adr-073-libp2p.md

```diff
@@ -0,0 +1,189 @@
+components, without breaking protocol or network-level
+interoperability and left the legacy connection/socket handling code.
+
+Following the release, the team has reexamined the state of the code
+and the design, as well as Tendermints requirements for. The notes
+from this process are available in the [P2P Roadmap
+RFC](../rfc/rfc-000-p2p.rst).
+
+While this ADR supersedes the decisions made in ADRs 60 and 61, it is
+also enabled by the completed portions of this work. Previously, the
+boundaries of peer management, message handling, and the higher level
+business logic (e.g. "the reactors") were intermingled, and core
+elements of the p2p system were responsible for the orchestration of
+the higher level business logic. The completed aspects of the refactor
+made it more obvious that the design implications of the legacy
+components had outsized influence on the entire implementation, that
+it would be difficult to iterate within the current abstractions, and
+it would not be viable to maintain interoperability with legacy
+systems while also achieving any of our broader objectives.
+
+LibP2P is a thoroughly specificed implementation of a peer-to-peer
+networking stack, designed specifically for systems such as
+ours. Adopting LibP2P as the basis of the Tendermint will allow the
+Tendermint team to focus more of their time on other differentiating
+aspects of the system, and make it possible for the ecosystem as a
+whole to take advantage of tooling and efforts of the LibP2P
+platform.
+
+## Alternative Approaches
+
+As discussed in the related RFC, the primary alternative is to
+continue development of tendermint's home grown peer-to-peer
+layer. While this gives the Tendermint team a maximal level of control
+over the peer system, the current peer system is unexceptional on its
+own merits, and the prospective maintenance burden for this system
+exceeds our tolerances for the medium term.
+
+It is also the case that Tendermint can and should differentiate
+itself not on the basis of its networking implementation or peer
+management tools, but on its consistent operator experience,
+battletested consensus algorithm, and ergonomic user experience.
+
+## Decision
+
+Tendermint will adopt libp2p during the 0.37 development cycle,
+replacing the bespoke Tendermint P2P stack. This will remove the
+`Endpoint`, `Transport`, `Connection`, and `PeerManager` abstractions
+and leave the reactors, `p2p.Router` and `p2p.Channel`
+abstractions.
+
+LibP2P may obviate the need for a dedicated peer exchange (PEX)
+reactor, which would also in turn obviate the need for a dedicated
+seed mode. If this is the case, then all of this functionality would
+be removed.
+
+If it turns out, based on the advice of Protocol Labs, that it makes
+sense to maintain separate pubsub or gossipsub topics
+per-message-type, then the `Router` abstraction may dissolve
+entirely.
+
+## Detailed Design
+
+### Implementation Changes
```

Comment:
will we kill our homecooked handshake protocol in favour of noise? 

This would close a couple security issues we have with our secret handshake.

Reply:
Yes, the handshake protocol will go away (and I think is implied by "replacing the connection," though I can call it out specifically.

Reply:
Is there a specific secure connection/identity mechanism in libp2p that can easily replace our handshake?

Reply:
I believe by default libp2p uses conventional TLS handshakes, at least for TCP. I'm less clear about QUIC, but I think it might also apply there. There are hooks for a custom handshake, though.